src/util/bip341.rs

```diff
@@ -0,0 +1,556 @@
+            Error::PrevoutsSize => write!(f, "Number of supplied prevouts differs from the number of inputs in transaction"),
+            Error::PrevoutIndex => write!(f, "The index requested is greater than available prevouts or different from the provided [Provided::Anyone] index"),
+            Error::PrevoutKind => write!(f, "A single prevout has been provided but all prevouts are needed without `ANYONECANPAY`")
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl Prevouts {
+    fn get_all(&self) -> Result<&[TxOut], Error> {
+        match self {
+            Prevouts::All(ref prevouts) => Ok(prevouts),
+            _ => Err(Error::PrevoutKind),
+        }
+    }
+
+    fn get(&self, input_index: usize) -> Result<&TxOut, Error> {
+        match self {
+            Prevouts::Anyone(index, prevout) => {
+                if input_index == *index {
+                    Ok(prevout)
+                } else {
+                    Err(Error::PrevoutIndex)
+                }
+            }
+            Prevouts::All(prevouts) => prevouts.get(input_index).ok_or(Error::PrevoutIndex),
+        }
+    }
+}
+
+/// A replacement for SigHashComponents which supports all sighash modes
+pub struct SigHashCache<'a> {
+    /// Access to transaction required for various introspection
+    tx: &'a Transaction,
+
+    /// Hash of all the previous outputs, computed as required
+    hash_prevouts: Option<sha256::Hash>,
+    /// Hash of all the input sequence nos, computed as required
+    hash_sequence: Option<sha256::Hash>,
+    /// Hash of all the outputs in this transaction, computed as required
+    hash_outputs: Option<sha256::Hash>,
+
+    /// Hash of all the prevout amounts, computed as required
+    hash_amounts: Option<sha256::Hash>,
+    /// Hash of all the prevout scriptpubkeys, computed as required
+    hash_scriptpubkeys: Option<sha256::Hash>,
+}
+
+impl<'a> SigHashCache<'a> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: &'a Transaction) -> Self {
+        SigHashCache {
+            tx,
+            hash_prevouts: None,
+            hash_sequence: None,
+            hash_outputs: None,
+            hash_amounts: None,
+            hash_scriptpubkeys: None,
+        }
+    }
+
+    /// Encode the BIP341 signing data for any flag type into a given object implementing a
+    /// std::io::Write trait.
+    pub fn encode_signing_data_to<Write: io::Write>(
+        &mut self,
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        annex: Option<Vec<u8>>,
+        script_path: Option<ScriptPath>,
+        sighash_type: SigHashType,
+    ) -> Result<(), Error> {
+        if let Prevouts::All(prevouts) = prevouts {
+            if prevouts.len() != self.tx.input.len() {
```

Comment:
We should also check `input_index < tx.input.len()`

Reply:
That one should be covered by the `Prevouts::get` method, isn't it?

Reply:
I am suggesting a check for `let txin = &self.tx.input[input_index];` not for prevouts

Reply:
added a new Error variant and the check in 00c5cf3